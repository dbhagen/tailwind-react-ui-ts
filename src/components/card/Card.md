Standard card functionality made up of:

- `<Card />` &mdash; wrapping component with overflow hidden to allow for full size image headers
- `<CardBody />` &mdash; padded element for main content of the card
- `<CardFooter />` &mdash; right aligns, reverses order & squares corners of elements to have CTAs in priority tab order

### Simple Example

```jsx
<Card>
  <CardBody>
    <p>Hello World</p>
  </CardBody>
</Card>
```

### Full Example

```jsx
<Card border shadow maxW="sm">
  <Image
    src="https://placekitten.com/1280/960"
    alt="Example image"
    aspectRatio={1280 / 960}
  />
  <CardBody>Hello World</CardBody>
  <CardFooter>
    <FillButton brand="primary">Read</FillButton>
    <OutlineButton brand="primary">Bookmark</OutlineButton>
  </CardFooter>
</Card>
```
