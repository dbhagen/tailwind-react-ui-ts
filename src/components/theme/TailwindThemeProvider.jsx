import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
