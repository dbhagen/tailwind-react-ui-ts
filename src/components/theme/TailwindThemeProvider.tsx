import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

// TODO: Remove Any Types
const TailwindThemeProvider = ({ theme, children }: { theme: any, children: any }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
