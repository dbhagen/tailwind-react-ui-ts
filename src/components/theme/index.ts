export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
