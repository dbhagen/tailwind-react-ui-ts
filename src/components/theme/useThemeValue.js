import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
