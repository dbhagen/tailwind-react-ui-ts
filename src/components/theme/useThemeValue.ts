import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, values: Object, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(undefined, values, prefix)
}
