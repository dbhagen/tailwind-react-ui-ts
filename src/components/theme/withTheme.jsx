import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
