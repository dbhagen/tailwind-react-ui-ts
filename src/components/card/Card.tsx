import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Box } from '../primitives'

const Card = ({ is, children, theme, surface, ...rest }) => (
  <Box
    is={is}
    overflow="hidden"
    rounded={theme.radius}
    bg={theme.surfaceColors[surface] || theme.brandColors[surface]}
    text={surface !== 'default' ? theme.textColors.on[surface] : undefined}
    {...rest}
  >
    {children}
  </Box>
)

Card.propTypes = {
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  surface: PropTypes.string,
}

Card.defaultProps = {
  is: 'section',
  children: undefined,
  surface: 'default',
}

export { Card as component }
export default withTheme(Card)
