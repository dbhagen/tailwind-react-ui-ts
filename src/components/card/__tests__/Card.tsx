import React from 'react'
import { shallow } from 'enzyme'

import { component as Card } from '../Card'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = shallow(<Card {...props}>Hello World!</Card>)

  return {
    props,
    wrapper,
  }
}

describe('Card', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
