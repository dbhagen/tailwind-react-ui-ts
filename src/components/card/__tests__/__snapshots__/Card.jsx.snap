// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders matching snapshot 1`] = `
<Box
  bg="white"
  inline={false}
  inlineBlock={false}
  is="section"
  overflow="hidden"
  rounded="rounded"
>
  Hello World!
</Box>
`;
